tools/bazar/controllers/ListController.php

```diff
@@ -0,0 +1,108 @@
+
+        if (isset($_POST['imported-list'])) {
+            foreach ($_POST['imported-list'] as $listRaw) {
+                $list = json_decode($listRaw, true);
+                $listManager->create($list['titre_liste'], $list['label']);
+            }
+            echo '<div class="alert alert-success">'._t('BAZ_LIST_IMPORT_SUCCESSFULL').'.</div>';
+        }
+
+        $lists = $listManager->getAll();
+
+        $values = [];
+        foreach ($lists as $key => $list) {
+            $values[$key]['title'] = $list['titre_liste'];
+            $values[$key]['options'] = $list['label'];
+            $values[$key]['canEdit'] = $GLOBALS['wiki']->HasAccess('write', $key);
+            $values[$key]['canDelete'] = $GLOBALS['wiki']->UserIsAdmin() || $GLOBALS['wiki']->UserIsOwner($key);
+        }
+
+        return $this->render('@bazar/lists/list_table.twig', [ 'lists' => $values ]);
+    }
+
+    public function create() 
+    {
+        $listManager = $this->getService(ListManager::class);
+
+        if( isset($_POST['valider']) ) {
+            $i = 1;
+            $values = [];
+            foreach ($_POST['label'] as $label) {
+                if (($label != null || $label != '') && ($_POST['id'][$i] != null || $_POST['id'][$i] != '')) {
+                    $values[$_POST['id'][$i]] = $label;
+                    ++$i;
+                }
+            }
+            
+            $listManager->create($_POST['titre_liste'], $values);
+
+            $GLOBALS['wiki']->Redirect(
+                $GLOBALS['wiki']->Href('', '', [BAZ_VARIABLE_VOIR => BAZ_VOIR_LISTES], false)
+            );
+        }
+
+        return $this->render('@bazar/lists/list_form.twig');
+    }
+
+    public function update($id) 
+    {
+        $listManager = $this->getService(ListManager::class);
+        $list = $listManager->getOne($id);
+
+        if( isset($_POST['valider']) ) {
+            if( $GLOBALS['wiki']->HasAccess('write', $id) ) {
+                $i = 1;
+                $values = [];
+
+                foreach ($_POST['label'] as $label) {
+                    if (($label != null || $label != '') && ($_POST['id'][$i] != null || $_POST['id'][$i] != '')) {
+                        $values[$_POST['id'][$i]] = $label;
+                    }
+                    ++$i;
+                }
+
+                $listManager->update($id, $_POST['titre_liste'], $values);
+
+                $GLOBALS['wiki']->Redirect(
+                    $GLOBALS['wiki']->Href('', '', [BAZ_VARIABLE_VOIR => BAZ_VOIR_LISTES], false)
+                );
+            } else {
+                throw new \Exception('Not allowed');
+            }
+        }
+
+        return $this->render('@bazar/lists/list_form.twig', [
+            'listId' => $id,
+            'title' => $list['titre_liste'],
+            'labels' => $list['label']
+        ]);
+    }
+    
+    public function delete($id)
+    {
+        $listManager = $this->getService(ListManager::class);
```

Comment:
puisque c'est appelé dans chaque méthode, faudrait il l'avoir comme attribut de la classe (et donc pouvoir faire des `$this->listManager` dans les méthodes) ?

Idem pour `$this->wiki` plutot que les $GLOBALS['wiki'] ?

Reply:
Bien vu pour `$GLOBALS['wiki']`, j'ai remplacé par `$this->wiki` dans tous les controlleurs.
`$this->wiki` est également disponible dans toutes les actions / handlers / formatters.

Pour listManager, je ne pense pas que c'est une bonne idée de le charger pour tous les controlleurs, car il va être utilisé principalement dans ListController. Je l'ai en revanche chargé dans le constructeur de ListController, puisqu'il est utilisé dans presque toutes les méthodes. (Comme les controlleurs sont des services, c'est possible de faire ça)